### ISSUE TITLE
Define how the current working directory should be handled when the manifest is within a sub-folder

### ISSUE THREAD

**Issue by ifbyol** [okteto/okteto#2855]
**Is your feature request related to a problem? Please describe.**
Some time ago there was change to execute the commands defined in the Okteto manifest from the directory where the manifest is stored instead of executing it from the directory where the `okteto deploy` command is being executed.

For example, imagine I'm in a folder called `root` and I execute `okteto deploy -f subfolder/okteto.yml`

This generated a couple of side effects that we would need to clarify and review:
* The filename stored in the internal config map is not the right one. We are storing only the filename instead of the relative path specified from the user when the command is executed. In the example above it means that we are storing `okteto.yml` instead of `subfolder/okteto.yml`. We have created a ticket for it. https://github.com/okteto/okteto/issues/2846
* The repo URL and branch calculation is now done from the directory where the manifest is stored instead of doing it from the directory where the `okteto deploy` command is executed. In the example above it means that we try to look for a `.git` folder within `subfolder` instead of doing it within `root` We have created an issue for it. https://github.com/okteto/okteto/issues/2849
* The name for the dev environment when it is not specified in the manifest is now being inferred from the folder where the manifest is stored instead of the folder where the `okteto deploy` command is executed. In the example above, we would name the dev environment `subfolder` instead of `root`. This probably make sense.

In other words, it seems we have changed the working directory be always the one where the manifest is stored.

There might be another secondary effects we didn't face yet

**Comment by pchico83** [comment]
When running `okteto deploy` there are two options:
a) the current working directory doesn't belong to a git repo. We can check this by checking if the current working directory or any of its parents has a `.git` folder on it
b) the current working directory belongs to a git repo

For a) we shouldn't store any of the following fields: `filename`, `branch` and `repoURL`.  These fields are for the Okteto UI to be able to show this information, and enable redeploys. But if we are not in a git repo, there is no need to store this information.

For b) we should get the `.git` folder and extract the git URL and branch from it. Also, `filename` should be the relative path of the okteto manifest with respect to the git root folder. Note that this value is a value independent from the current working directory. `filename` should only be stored for okteto manifests or compose files. For automatic pipelines, @ifbyol not sure if we could store the folder

**Comment by ifbyol** [comment]
@pchico83 Yes, we can store the folder. In fact, we were storing the folder but with the changes to use current working directory we don't store folders anymore.

If the user specifies a folder with the `-f` flag, I think we should store that

**Comment by pchico83** [comment]
@ifbyol in that case, when using the automatic pipeline, we should store the current working directory as the `filename` no? Specially if it's different than the root folder. Otherwise, the automatic pipeline will be broken when redeploying from the UI

**Comment by ifbyol** [comment]
@pchico83  In that case, I think current working directory should be the same than the specified folder with `-f` (being relative to the git repo). I cannot think in a case in which they would be different.

If the filename/folder specified with `-f` don't exist, I think we should behave in the same way in both cases (directly fail or fallback to autodiscoverability in the root folder of the repo)

**Comment by ifbyol** [comment]
@pchico83 Also, one scenario to bear in mind, maybe a corner case but we have to think at least if we want to support it or not. A git repo within another repo, I guess we should store the first git repo we find?

**Comment by pchico83** [comment]
@ifbyol yes, let's store the first parent. I think it's a corner case, I would go with simplicity for now

**Comment by jLopezbarb** [comment]
I think this issue would be solved by deprecating the use of `os.Chdir()` and `os.Getwd()`, by making the code not rely on it to parse the files/discover where the `.git` file is. The following two comments describe the proposed solution for each problem.

**Comment by jLopezbarb** [comment]
The first problem is the change of the working directory to parse the files with the relative paths, this can produce problems when generating absolute paths in other parts of the code rather than the parsing, I think the best thing is to create a new data structure that collects the relative path that the user has defined on the manifest and an absolute one (given by the user or calculated by us when we set the defaults). This way we could always use the absolute path which is more accurate than the relative path and still have the relative one for those use cases where we need a relative path. For example, the volume structure in the docker-compose parser would look like this:

```yaml
services:
  volumes:
  - api:/usr
```

```golang
type volume struct {
  localPath struct {
    relativePath string // api
    absolutePath string // /Users/src/api
  }
  remotePath string     // /usr
}
```

The fields that we would need to change to support this new type would be:

| dev                    | buildInfo  | deploy           | docker-compose/stack       |
| :----------------------- | ------------ | ------------------ | ---------------------------- |
| volume.Localpath       | context    | compose.manifest | services.env_files         |
| sync.LocalPath         | dockerfile |                  | services.volumes.localPath |
| sync.folders.LocalPath |            |                  |                            |
| secrets.LocalPath      |            |                  |                            |
| envfiles               |            |                  |                            |

The following scenarios should work the same(deploy is an example, the same should happen with the rest of commands like up/down/exec...):
- File on the same path: `okteto deploy -f manifest.yml`.
- File on the previous path: `okteto deploy -f ../manifest.yml`.
- File on a child directory: `okteto deploy -f child/manifest.yml`. 
- Parsing from discovery: `okteto deploy`.
- Manifest on `.okteto` folder: `okteto deploy -f .okteto/manifest.yml`.

**Comment by jLopezbarb** [comment]
The second problem is figuring out which `.git` to use and how to reference the filename in the UI for its possible redeployment. We should only store the filename if the user provides it, or if the path where the `.git` file is different to the one that the command is executed. In this second case, I think it is best to explain each of the use cases and how it should behave:

1. `.git` on the same folder as the manifest:
We need to extract the repository and branch from the `.git` repo that is at the same level as the okteto manifest.

```
├── (executed command here)
├── .git (repo and branch extracted from here)
├── okteto.yml
```

```
filename -> okteto.yml (if command has -f okteto.yml or empty if there is no file flag)
branch -> The one from .git
repoURL -> The one from .git
```

2. `.git` on a previous folder than the manifest but executing from the folder of `.git`:

We need to extract the repository and branch from the `.git` repo where the command is executed.

```
├── (executed command here)
├── .git (repo and branch extracted from here)
├── manifest
│   ├── okteto.yml
```

```
filename -> manifest/okteto.yml (if command has -f okteto.yml or empty if there is no file flag)
branch -> The one from .git
repoURL -> The one from .git
```

3. `.git` folder doesn't exist on the execution folder and a child:

We don't extract any repository or branch because there is no `.git` folder where we can extract the information. We don't have to calculate what would be the filename because there is no point on calculating it if the user is not going to be able to redeploy from the UI

```
├── (executed command here)
├── here
├── manifest
│   ├── another-here
│   ├── okteto.yml
```

```
filename -> none
branch -> none
repoURL -> none
```

4. `.git` folder exists on the same folder as manifest but executes from a previous folder without `.git`:

We can store the repository and branch from the .git in the manifest folder, since this case could be an user executing from its home a repo in an specific folder. In this case, since we are storing the `.git` that is at the root level, we have to calculate the filename from the `.git` folder instead of from the command execution path. This can be a common use case if the user runs `okteto deploy` from `$HOME`. 

```
├── (executed command here)
├── here
├── manifest
│   ├── .git (repo and branch extracted from here)
│   ├── okteto.yml
```

```
filename -> okteto.yml
branch -> The one from .git
repoURL -> The one from .git
```

5. `.git` folder exists on the same folder as manifest but executes from a previous folder with `.git`:

We should be storing the first `.git` folder that we find. In this case the one from the command we execute.

```
├── (executed command here)
├── .git (repo and branch extracted from here)
├── manifest
│   ├── .git 
│   ├── okteto.yml
```

```
filename -> manifest/okteto.yml
branch -> The one from the first .git
repoURL -> The one from the first .git
```

6. `.git` doesn't exists on the execution folder and on a child, but exists on a previous folder:

We should store the `.git` and calculate the paths from the `.git` folder  
```
├── ../.git (repo and branch extracted from here)
├── (executed command here)
├── here
├── manifest
│   ├── another-here
│   ├── okteto.yml
```

```
filename -> folder_name/manifest/okteto.yml
branch -> The one from the previus .git
repoURL -> The one from the previus .git
```

7. `.git` exists on a previous folder and on a child folder:

We should store the `.git` and calculate the paths from the `.git` folder that is on the previous path.

```
├── ../.git (repo and branch extracted from here)
├── (executed command here)
├── here
├── manifest
│   ├── .git
│   ├── another-here
│   ├── okteto.yml
```

```
filename -> folder_name/manifest/okteto.yml
branch -> The one from the previus .git
repoURL -> The one from the previus .git
```

These scenarios follows the following rules to discover .git:
1. Is `.git` on same folder as execution?
2. Is `.git` on same folder as `-f` flag? If yes and `-f` file on different folder than execution folder recalculate filename path
3. Is `.git` on previous folder? If yes recalculate filename path
4. Is `.git` on a child folder? If yes recalculate finame path from child folder

**Comment by pchico83** [comment]
@jLopezbarb for `.git` problem we should get the first parent `.git` folder of the okteto manifest file.
If there isn't a `.git` folder, don't store either the repo URL or `filename` (this info is needed to redeploy a git repo from the UI. if it's not a git repo, we don't need it). Otherwise:
- Use that `.git` folder to get the repo URL
- if `-f` is passed as an argument, store the relative path vs `.git` as the `filename`
- if `-f` isn't passed as an argument, and `cwd` is the same as the location of `.git` folder, don't store `filename`
- if `-f` isn't passed as an argument, and `cwd` isn't the same as the location of `.git` folder, store the relative path of `cwd` versus `.fit` as the `filename`. For example:
```
├── .git (repo and branch extracted from here)
├── frontend
│   ├── execute okteto deploy here
│   ├── okteto.yml
```
`filename` world be `frontend`, because from the root of the repo, this equivalent to run `okteto deploy -f frontend`

**Comment by ifbyol** [comment]
To be considered defining this. Currently, if you execute `okteto deploy -f` with a non-existent file, we detect that the specified file doesn't exist and we keep executing it with the autodiscoverability feature. But we don't display any message to the user.

We should define the behavior we want when the user specifies a file that doesn't exist. If the current behavior is the one we want, we might evaluate to show a warning o similar to let them know that the file doesn't exist and we will try to find another one to use

**Comment by pchico83** [comment]
@jLopezbarb do we have a final spec for this? is it ready for implementation?

**Comment by jLopezbarb** [comment]
There are two specs in notion already (one functional and another one technical). We will need some refinement before working on it but it should be mostly moving content from the technical
